server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java

```diff
@@ -53,17 +54,75 @@ public static void validateMapperStructure(Collection<ObjectMapper> objectMapper
         }
 
         Set<String> fieldNames = new HashSet<>();
-        Stream.concat(fieldMappers.stream(), fieldAliasMappers.stream())
-            .forEach(mapper -> {
-                String name = mapper.name();
-                if (objectFullNames.contains(name)) {
-                    throw new IllegalArgumentException("Field [" + name + "] is defined both as an object and a field.");
-                } else if (fieldNames.add(name) == false) {
-                    throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
-                }
-            });
+        for (FieldMapper fieldMapper : fieldMappers) {
+            String name = fieldMapper.name();
+            if (objectFullNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an object and a field.");
+            } else if (fieldNames.add(name) == false) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
+            }
+
+            validateFieldMapper(fieldMapper, fieldTypes);
+        }
+
+        Set<String> fieldAliasNames = new HashSet<>();
+        for (FieldAliasMapper fieldAliasMapper : fieldAliasMappers) {
+            String name = fieldAliasMapper.name();
+            if (objectFullNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an object and a field.");
+            } else if (fieldNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an alias and a concrete field.");
+            } else if (fieldAliasNames.add(name) == false) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
+            }
+
+            validateFieldAliasMapper(name, fieldAliasMapper.path(), fieldNames, fieldAliasNames);
+        }
     }
 
+    /**
+     * Checks that the new field mapper does not conflict with existing mappings.
+     */
+    private static void validateFieldMapper(FieldMapper fieldMapper,
+                                            FieldTypeLookup fieldTypes) {
+        MappedFieldType newFieldType = fieldMapper.fieldType();
+        MappedFieldType existingFieldType = fieldTypes.get(newFieldType.name());
+
+        if (existingFieldType != null && Objects.equals(newFieldType, existingFieldType) == false) {
+            List<String> conflicts = new ArrayList<>();
+            existingFieldType.checkCompatibility(newFieldType, conflicts);
+            if (conflicts.isEmpty() == false) {
+                throw new IllegalArgumentException("Mapper for [" + newFieldType.name() +
+                    "] conflicts with existing mapping:\n" + conflicts.toString());
+            }
+        }
```

Comment:
Before we were also checking that a new field name is not used to define a field alias. Is there any reason to remove this check now?

Reply:
This check still exists, but it's moved into the main method `validateNewMappers`:

```
else if (fieldNames.contains(name)) {
    throw new IllegalArgumentException("Field [" + name + "] is defined both as an alias and a concrete field.");
```

I thought it was clearer to move it, because this is where we check for other conflicts (like the same name being used for an object and a field).